Source/DafnyCore/Verifier/Translator.ExpressionWellformed.cs

```diff
@@ -916,7 +916,7 @@ private class WFOptions {
                     zero = Bpl.Expr.Literal(0);
                   }
                   CheckWellformed(e.E1, wfOptions, locals, builder, etran);
-                  builder.Add(Assert(GetToken(expr), Bpl.Expr.Neq(etran.TrExpr(e.E1), zero), new PODesc.DivisorNonZero(), wfOptions.AssertKv));
+                  builder.Add(Assert(GetToken(expr), Bpl.Expr.Neq(etran.TrExpr(e.E1), zero), new PODesc.DivisorNonZero(e.E1), wfOptions.AssertKv));
```

Comment:
My long-term wish would be that we could construct the expression once and then put that expression both in an assertion and in the proof obligation description. That way, we'd know that the two expressions always match. It's probably not worth holding this PR up over that, if only because the current side conditions are expressed as Boogie expressions and we really want Dafny expressions to show to the user. But it would be lovely to eventually be in a state where that was possible.

Reply:
I was troubled by this as well. Is there any way we could test (or even prove) that the Dafny-level expression is encoded as a Boogie expression that is equivalent to the actual Boogie expression?

Reply:
For now, whenever it was built already, I will ensure I'm reusing the same expression that is first translated to Boogie. That said, proving is out of control is our code is not written in Dafny :-) Regarding testing, we cannot necessarily execute the code.
We could however emit an internal assertion in the code that the Dafny expression is equivalent to the Boogie expression.

However, this is not practical/desirable for many reasons
- **Not practical to generate the expression at translation level** Code actions will generate the assertion expression on demand on failed assertions, so that we avoid generating 2x more expressions. At this point, the verifier has already finished to run.
- **Not practical to always create a Dafny expression first** What we send to Boogie is often might not be not practical for the user. For example, for termination checks, we send a O(n^2) expression which is practical for Boogie to prove; but for the user, we want a linear nested expression
- **Not necessary because safeguards:** This expression will only be served as an hint. If it's too weak, anyway the user will see that it won't help proving the assertion. If it's too strong because we can't expose some Boogie concepts in Dafny, then not only we can't prove equivalence, but the user won't be able to prove it either, so the verifier acts as a safeguard
 in Boogie.
- **Not meaningful for many assertions** For `assert A == B`, we will want to generate `calc {A; B;}.` It does not make sense to generate the calc statement and convert it to Boogie Down the road, we will generate more than simple expressions, we will generate whole code actions, e.g. to prove decreases clauses in while loop or prove frame conditions.